### PR TITLE
Merkle Search Tree

### PR DESCRIPTION
**Pull request by dholms**
Implementing Merkle Search Tree data structure for repo layer.

Described here: https://hal.inria.fr/hal-02303490/document

This is an immutable implementation. So when using the tree:
```ts
// do this
let mst = await MST.create(...)
mst = await mst.add(key, cid)
await mst.save() // persists all changes to the blockstore

// don't do this
const mst = await MST.create(...)
await mst.add(key, cid)
```

The tree exposes APIs to `add`, `edit`, and `delete` records. As well as to `diff` two different trees.

A couple notes on CBOR encoding:

There are _never_ two neighboring subtrees (if there are, they should be merged). Therefore, we can represent a node as an array of leaves & pointers to their right neighbor (possibly null). Along with a pointer to the left-most subtree (also possibly null)

_Most_ leaves in a subtree will have overlap (for instance, two posts created around the same time in the same collection). We do compression on prefixes by describing keys as:
- the length of the prefix that it shares in common with the preceding key
- the rest of the string

For example:
If the first leaf in a tree is `bsky/posts/abcdefg` and the second is `bsky/posts/abcdehi`
Then the first will be described as `prefix: 0, key: 'bsky/posts/abcdefg'` and teh second will be described as `prefix: 16, key: 'hi'`.

We get the following CBOR encoding
```
{
  l: CID | null // a (possibly null) pointer to the left most subtree
  e: {
    p: number, // prefix length that this leaf shares in common with the previous leaf
    k: string, // the rest of the key
    v: CID, // the leaf value
    t: CID | null // a (possibly null) pointer to subtree to the immediate right of this leaf
  }[]
}
```

### CODE REVIEW

File: packages/common/src/repo/mst/diff.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+export type MstAdd = {
+  key: string
+  cid: CID
+}
+
+export type MstUpdate = {
+  key: string
+  old: CID
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/common/src/repo/mst/diff.ts:73]: I like `prev` 👍

File: packages/common/src/repo/mst/mst.ts [hunk 2]
```diff
@@ -0,0 +1,681 @@
...
+import * as uint8arrays from 'uint8arrays'
+import IpldStore from '../../blockstore/ipld-store'
+import { sha256 } from '@adxp/crypto'
+
+import z from 'zod'
+import { schema } from '../../common/types'
+import * as check from '../../common/check'
+import { MstDiff } from './diff'
+
+const subTreePointer = z.union([schema.cid, z.null()])
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/common/src/repo/mst/mst.ts:44]: ah forgot about that. i'll switch it out

File: packages/common/src/repo/mst/mst.ts [hunk 3]
```diff
@@ -0,0 +1,681 @@
...
+      return this.key === entry.key && this.value.equals(entry.value)
+    } else {
+      return false
+    }
+  }
+}
+
+export const leadingZerosOnHash = async (key: string): Promise<number> => {
+  const hash = await sha256(key)
+  const b32 = uint8arrays.toString(hash, 'base32')
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/common/src/repo/mst/mst.ts:598]: ah yup meant to fix this up & make it a param on the class

File: packages/common/src/repo/mst/mst.ts [hunk 4]
```diff
@@ -0,0 +1,681 @@
...
+  }
+
+  // Matching Leaf interface
+  // -------------------
+
+  isTree(): this is MST {
+    return true
+  }
+
+  isLeaf(): this is Leaf {
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/common/src/repo/mst/mst.ts:563]: needed the runtime type guard `is Leaf` which you can't do with just a bool :/

File: packages/common/src/repo/mst/mst.ts [hunk 5]
```diff
@@ -0,0 +1,681 @@
...
+        await MST.fromCid(blockstore, entry.t, layer ? layer - 1 : undefined),
+      )
+    }
+  }
+  return entries
+}
+
+const serializeNodeData = (entries: NodeEntry[]): NodeData => {
+  const data: NodeData = {
+    l: null,
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/common/src/repo/mst/mst.ts:646]: well I intentionally used single chars to cut down on the space these objects took up when encoded to CBOR.

What're your thoughts on object size v legibility? (or is there some other CBOR trick I'm not thinking of?)

File: packages/common/src/repo/mst/mst.ts [hunk 6]
```diff
@@ -0,0 +1,681 @@
...
+    if (next?.isTree()) {
+      subtree = next.pointer
+      i++
+    }
+    const prefixLen = countPrefixLen(lastKey, leaf.key)
+    data.e.push({
+      p: prefixLen,
+      k: leaf.key.slice(prefixLen),
+      v: leaf.value,
+      t: subtree,
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/common/src/repo/mst/mst.ts:672]: I also prefer that. But this goes back to the problem of these things need to be deserialized in any language. and I think a heterogenous array doesn't work in some languages (golang)

File: packages/common/src/repo/mst/mst.ts [hunk 7]
```diff
@@ -0,0 +1,681 @@
...
+    if (keyZeros === layer) {
+      // it belongs in this layer
+      const index = await this.findGtOrEqualLeafIndex(key)
+      const found = await this.atIndex(index)
+      if (found && found.equals(newLeaf)) {
+        throw new Error(`There is already a value at key: ${key}`)
+      }
+      const prevNode = await this.atIndex(index - 1)
+      if (!prevNode || prevNode.isLeaf()) {
+        // if entry before is a leaf, (or we're on far left) we can just splice in
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/common/src/repo/mst/mst.ts:176]: `!prevNode`

If the prevNode does not exist then we're on the far left

File: packages/common/src/repo/mst/mst.ts [hunk 8]
```diff
@@ -0,0 +1,681 @@
...
+import { sha256 } from '@adxp/crypto'
+
+import z from 'zod'
+import { schema } from '../../common/types'
+import * as check from '../../common/check'
+import { MstDiff } from './diff'
+
+const subTreePointer = z.union([schema.cid, z.null()])
+const treeEntry = z.object({
+  p: z.number(), // prefix count that is the same as the prev key
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/common/src/repo/mst/mst.ts:46]: Yeah I vote we allow only utf8 strings. & counts characters

[the remark this answers is not in the document]
Comment (AaronGoldman) [reply packages/common/src/repo/mst/mst.ts]: yup, looks like it is the count of unicode characters (codepoints) not bytes.
we can just say that it is the count of characters so it is clear it is not bytes 
```ts
export const countPrefixLen = (a: string, b: string): number => {
  let i
  for (i = 0; i < a.length; i++) { if (a[i] !== b[i]) {break} }
  return i
}
```

File: packages/common/src/repo/mst/mst.ts [hunk 9]
```diff
@@ -0,0 +1,755 @@
...
+type SupportedBases = 'base2' | 'base8' | 'base16' | 'base32' | 'base64'
+
+export const leadingZerosOnHash = async (
+  key: string,
+  fanout: Fanout,
+): Promise<number> => {
+  if ([2, 8, 16, 32, 64].indexOf(fanout) < 0) {
+    throw new Error(`Not a valid fanout: ${fanout}`)
+  }
+  const base: SupportedBases = `base${fanout}`
```
Comment (AaronGoldman) [inline packages/common/src/repo/mst/mst.ts:632]: may be cleaner to define a `zeroChar`
```ts
  if (![2, 8, 16, 32, 64].includes(fanout)){
    throw new Error(`Not a valid fanout: ${fanout}`)
  }
  const base: SupportedBases = `base${fanout}`
  const zeroChar = uint8arrays.toString(new Uint8Array(1), base)[0] // first char of encoded 0
  const hash = await sha256(key)
  const encoded = uint8arrays.toString(hash, base)
  let i
  for (i = 0; i < a.length; i++) {
    if (encoded[i] !== zeroChar) {
      break
    }
  }
  return i
```

File: packages/common/src/repo/mst/mst.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (AaronGoldman) [reply packages/common/src/repo/mst/mst.ts:697]: I don't know if CBOR interns strings let me check.
If they dictionary encode the keys the it is the cost of storing the key string once if they don't then the keys will be in every entry 64ish times :(

File: packages/common/src/repo/mst/mst.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (AaronGoldman) [reply packages/common/src/repo/mst/mst.ts:723]: fair

File: packages/common/src/repo/mst/mst.ts (hunk 9 above)
Comment (dholms) [reply packages/common/src/repo/mst/mst.ts:678]: oh i like that

File: packages/common/src/repo/mst/mst.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (AaronGoldman) [reply packages/common/src/repo/mst/mst.ts:736]: ok I am convinced even in CBOR the single char name save space we can make up for it with good documentation.

**Pull request merged**